increase resilience to rate limiting QUALITY-2043

cprovencher opened the pull request:
1. changed default log level to debug so it is visible in all of CI and helps us with dcos-launch tickets
2. added a check for get_on_prem_cluster  https://jira.mesosphere.com/browse/QUALITY-2043

cprovencher committed:
increase resilience to rate limiting QUALITY-2043

sschneid commented on dcos_launch/aws.py:
```diff
@@ -185,3 +187,39 @@ def get_cluster_hosts(self):
+        # hosts will not be found inside the stack. So here we implement a retrying logic with exponential backoff that
+        # asserts all the cluster hosts are found in the stack.
+        sleep_time = 2
+        while True:
```
Maybe have some sort of `max_retry` to keep it from looping forever?

sschneid commented:
We're currently requiring the [`retrying`](https://github.com/dcos/dcos-launch/blob/master/requirements.txt#L18) library, could this be used instead of constructing retry logic?

cprovencher commented:
@sschneid  I just added a maximum and also implemented the real exp backoff (random between 0 and 2^tries - 1) instead of what I had done (wrong as sleep time rises way too fast)

I would argue against using retrying in this case because it's cumbersome when you don't want to retry based on a result or an exception, but instead want to check 5 conditions and do an exponential backoff

sschneid reviewed (approved):
lgtm!

margaret commented:
I think this would be how you could do it with the retrying lib, although I haven't tested it
```python
def get_onprem_cluster(self):
    num_masters = int(self.config['num_masters'])
    num_private_agents = int(self.config['num_private_agents'])
    num_public_agents = int(self.config['num_public_agents'])

    @retry(wait_exponential_multiplier=10*1000, wait_exponential_max=20*60*1000, retry_on_result=False)
    def _get_cluster():
        cluster = dcos_test_utils.onprem.OnpremCluster.from_hosts(
            bootstrap_host=self.get_bootstrap_host(),
            cluster_hosts=self.get_cluster_hosts(),
            num_masters=num_masters,
            num_private_agents=num_private_agents,
            num_public_agents=num_public_agents)

        num_masters_found = len(cluster.masters)
        num_private_agents_found = len(cluster.private_agents)
        num_public_agents_found = len(cluster.public_agents)
        bootstrap_host_found = bool(cluster.bootstrap_host)

        cluster_matches_config = num_masters_found == num_masters and num_private_agents_found == num_private_agents and \
            num_public_agents_found == num_public_agents and bootstrap_host_found

        if not cluster_matches_config:
            msg = ("Not all required hosts were found:\n"
                   "\tmaster count: expected {}, found {}\n"
                   "\tprivate agent count: expected {}, found {}\n"
                   "\tpublic agent count: expected {}, found {}\n"
                   "\tbootstrap host found: {}")
            msg = msg.format(num_masters, num_masters_found,
                             num_private_agents, num_private_agents_found,
                             num_public_agents, num_public_agents_found,
                             bootstrap_host_found)

            log.info("Stack not ready yet for DC/OS installation. " + msg)

        return cluster_matches_config

    return _get_cluster()
```
How did you test your implementation? You mentioned that an earlier one wasn't right. How'd you catch that?

cprovencher commented:
@margaret you're right - my cumbersome argument is wrong, however the retrying lib doesn't do exponential backoff correctly - sleep time rises way too fast. All it does is 2^x * a constant where x is the number of tries. The correct way exponential backoff is implemented is random between 0 and 2^x - 1 where x is the number of tries

The way I figured out my previous code was wrong is not the result of any testing - I just hadn't implemented exponential backoff. I was doing 2\*2=4   4\*4=16   8\*8=64 etc. which is even worse than 2^x

cprovencher committed:
Use retrying lib to fix rate limiting issue

cprovencher commented:
@margaret  you win, using retrying is more consistent with the rest of the project and more readable

cprovencher committed:
Replace platforms.aws rate limiting decorator with retrying.retry decorator
